panther-labs/panther
# Managed notifications multiple sources same bucket

## Background

The Panther notifications feature didn't really account for multiple s3 sources for the same bucket. This scenario
would make the feature unusable if some s3 sources would have the same bucket and overlapping prefixes.

Although this might not seem a common scenario, it is currently allowed and we should support it.

## Changes

- The code that configures bucket notifications now takes into account the s3 prefixes for all the saved sources with the same bucket.
- Don't save the created notifications or the created SNS topic in the db anymore. Assume that a notification is managed by Panther if its name starts with `panther-managed-`
- The healthcheck now becomes green if the prefix defined for a notification is a string prefix of the source's prefix (i.e. we don't require exact equality)

## Testing

- Preexisting unit tests
- Manual QA:
    - Create 3 sources with overlapping prefixes and check bucket notifications
    - Add more prefixes to a source, check bucket notifications
    - Delete the sources one by one, checking bucket notifications in the process

Closes https://app.asana.com/0/1199961111326839/1199961165787232
